### PR TITLE
sig-scheduling: Scheduler code hierarchy overview

### PR DESCRIPTION
**Pull request by ingvagabund**
Reflect the current state of implementation of the scheduler code

Based on https://docs.google.com/document/d/1t2QoASmg_BssFczl8n7UNTDWKgncMSrLTX802fQ0lK8/edit#

**Which issue(s) this PR fixes**:

Part of https://github.com/kubernetes/community/issues/5237

### CODE REVIEW

**Review by markyjackson-taulia — APPROVED** [review]
/lgtm

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+For every Pod that the scheduler discovers, the scheduler becomes responsible
+for finding the best Node for that Pod to run on.
+Scheduling in general is quite an extensive field in computer science which takes
+into account various range of constraints and limitations.
+Each workload may require a different approach to achieve optimal scheduling results.
+The kube-scheduler provided by Kubernetes project was constructed with a goal
+to provide high throughput at the cost of being simple.
+To help in building a scheduler (the default or a custom one) and to share
+elements of the scheduling logic,
+[the scheduling framework](https://www.google.com/url?q=https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/&sa=D&source=editors&ust=1617027676318000&usg=AOvVaw2vEsCZwet-BiQGhy0mVpZu)
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:15]: Simplify the URL:  https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+a fully functional unit. This document aims at describing how all the individual
+pieces are put together and what’s their role in the overall architecture
+so a developer can quickly orient in the code.
+
+## Scheduling a pod
+
+The default scheduler instance has a loop running indefinitely
+which (everytime there’s a pod) is responsible for invoking the scheduling logic
+and making sure a pod gets either a node assigned or requeued for future processing.
+Each loop consists of a scheduling and a binding cycle.
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:28]:
```suggestion
Each loop consists of a blocking scheduling and a non-blocking binding cycle.
```

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+The scheduling cycle is responsible for running the scheduling algorithm selecting
+the most suitable node for placing the pod.
+The binding cycle makes sure the kube-apiserver is made aware of the selected
+node at the right time (e.g. gang scheduling requires to wait until all pods
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:32]: We may reword it like this:

```
A pod may be bound immediately, or in the case of gang scheduling,
wait until all its sibling pods have their node assigned.
```

We shouldn't word it as "binding all pods at once" as each binding cycle is a goroutine. A single binding goroutine can only cover its own pod.

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+1. Get the next pod for scheduling
+1. Schedule a pod with provided algorithm
+1. If a pod fails to be scheduled due to `FitError`, run preemption plugin in
+   `PostFilterPlugin` (if the plugin is registered) to nominate a node where
+   the pods can run. If preemption was successful,
+   let the current pod be aware of the nominated node.
+   Handle the error, get the next pod and start over.
+1. If the scheduling algorithm finds a suitable node, store the pod into
+   the scheduler cache (`AssumePod` operation) and run plugins from the `Reserve`
+   and `Permit` extension point in that order. In case either of the plugins fails,
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:47]: either -> any

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+### Binding cycle
+
+Consists of the following four extension points ran in the same order:
+- [WaitOnPermit](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L560)
+  (internal API): Some plugins from the Permit extension point
+  may send a request for an operation requiring to wait for a condition
+  (e.g. wait for additional resources to be available or wait for all pods
+  in a gang to be assumed).
+  The `WaitOnPermit` extension point allows you to wait for such a condition to be met.
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:61]: Generally `WaitOnPermit` is the 2nd part of `Permit` extension point, and works internally on the framework side. So maybe reword it a bit to avoid confusion as when users think of "extension point",  they usually think of it as an extendable interface, which is `Permit()`.

```
Under the hood, `WaitOnPermit` waits for such a condition to be met within a timeout threshold.
```

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+  the default scheduler codebase root directory
+- [pkg/scheduler/core](https://github.com/kubernetes/kubernetes/tree/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core):
+  location of the default scheduling algorithm
+- [pkg/scheduler/framework](https://github.com/kubernetes/kubernetes/tree/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/framework):
+  scheduling framework alongside plugins
+- [pkg/scheduler/internal](https://github.com/kubernetes/kubernetes/tree/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/internal):
+  implementation of the cache, queues and other internal elements
+- [staging/src/k8s.io/kube-scheduler](https://github.com/kubernetes/kubernetes/tree/a651804427dd9a15bb91e1c4fb7a79994e4817a2/staging/src/k8s.io/kube-scheduler):
+  location of ComponentConfig API types
+
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:84]: Also mention the test codebase? (ut, integration, benchmark, e2e)

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+## Initial startup configuration
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:85]: I guess it should be H3 title?

```suggestion
### Initial startup configuration
```

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+- Scheduling profiles instantiated to tailor a framework for each profile
+  to better suit pod placement (each profile defines a set of plugins to use)
+- Handler functions for getting the next pod for scheduling (`NextPod`) and error handling (`Error`)
+
+The following steps are taken during the process of creating a scheduler instance:
+- Scheduler [cache is initialized](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L206)
+- Both in-tree and out-of-tree registries with plugins are [merged together](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L208-L211)
+- Metrics are [registered](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L232)
+- [Configurator](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L215-L230)
+  building a scheduler instances (wiring the cache, plugin registry,
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:151]: instances -> instance

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+  scheduling algorithm and other elements together)
+- Event handlers [are registered](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L273)
+  to allow the scheduler to react on changes in PVs,
+  PVCs, services and other objects relevant for scheduling (eventually,
+  each plugin will define a set of events on which it reacts,
+  see [kubernetes/kubernetes#94009](https://github.com/kubernetes/kubernetes/issues/94009)
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:157]: I have a new issue compiling completed and on-going items: https://github.com/kubernetes/kubernetes/issues/100347

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+  for more details).
+
+The following diagram shows how individual elements are connected together
+once initialized. Event handlers make sure pods are properly enqueued
+in the [scheduling queues](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-scheduling/scheduler_queues.md),
+the cache is updated with pods and nodes
+as they go (to provide up-to-date snapshot). Scheduling algorithm and the binding cycle
+have the right instances of the framework available (one instance of the framework per a profile).
+
+![Default scheduler architecture](default_scheduler_architecture.png "Default scheduling architecture")
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:167]: Is the extra wording "Default scheduling architecture" intentional?

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+#### Scheduling framework
+
+It’s code is currently located under `pkg/scheduler/framework`.
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:171]: It's -> Its

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+Used as building blocks for any scheduling algorithm.
+
+When a [plugin is initialized](https://github.com/kubernetes/kubernetes/blob/4740173f3378ef9d0dc59b0aa9299444a97d0818/pkg/scheduler/framework/runtime/framework.go#L310),
+it’s passed a [framework handler](https://github.com/kubernetes/kubernetes/blob/4740173f3378ef9d0dc59b0aa9299444a97d0818/pkg/scheduler/framework/runtime/framework.go#L251-L264)
+which provides interfaces to access and/or manipulate pods, nodes, clientset,
+event recorder and other handlers every plugin needs to implement its functionality.
+
+#### Scheduler cache
+
+Cache responsible for capturing the current state of a cluster.
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:183]:
```suggestion
Cache is responsible for capturing the current state of a cluster.
```

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+- `Extenders`: currently exposed only for testing
+
+Each cycle of the default algorithm implementation consists of:
+1. Taking the [current snapshot](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L101)
+   from the scheduling cache
+1. [Filter out all nodes](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L110)
+   not feasible for scheduling a pod
+   1. Run [PreFilter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L230)
+      first (preprocessing phase, e.g. computing pod [anti-]affinity relations)
+   1. Run [Filter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L261):
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:255]: Run Filter plugins in parallel:

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+      filter out all nodes which does not satisfy pod’s constraints
+      (e.g. sufficient resources, node affinity, etc.), including running filter extenders.
+1. In case there are at least two feasible nodes for scheduling, run [scoring plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L133):
+   1. Run [PreScore plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L427)
+      first (preprocessing phase)
+   1. Run [Score plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L433):
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:261]: Run Score plugins in parallel:

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+      and add it to the total score of each node
+1. [Select](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L138)
+   and [give back a node](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L141-L145)
+   with the highest score. If there’s only a [single feasible node](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L125-L131)
+   skip `PreScore`, `Score` and `NormalizeScore` extension points
+   and give back the node right away. If there’s no feasible node, report it.
+
+Be aware of:
+- [PostFilter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L479)
+are ran outside the scheduling algorithm.
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:278]: "ran outside the scheduling algorithm" sounds ambiguous to me. I'd suggest putting it into the regular flow, in step 2.iii above.

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md (hunk 10 above)
Comment (ingvagabund) [reply contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:167]: Probably not. I wanted to point out a developer can build a custom scheduler which can have a different architecture. The "Default" here refers to kube-scheduler.

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md (hunk 6 above)
Comment (ingvagabund) [reply contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:90]: Which unit tests do you have in mind?

**Comment by ingvagabund** [comment]
@Huang-Wei thanks for the review!!! Comments addressed leaving just one question about unit tests (please see above).

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md (hunk 6 above)
Comment (Huang-Wei) [reply contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:90]: I mean the regular ones colocated with the source code (*_test.go)

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+1. Taking the [current snapshot](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L101)
+   from the scheduling cache
+1. [Filter out all nodes](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L110)
+   not feasible for scheduling a pod
+   1. Run [PreFilter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L230)
+      first (preprocessing phase, e.g. computing pod [anti-]affinity relations)
+   1. Run [Filter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L261) in parallel:
+      filter out all nodes which does not satisfy pod’s constraints
+      (e.g. sufficient resources, node affinity, etc.), including running filter extenders.
+   1. Run [PostFilter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L479)
```
Comment (Huang-Wei) [inline contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:264]: Run PostFilter plugins if no node can fit the incoming pod.

**Review by Huang-Wei — COMMENTED** [review]
A nit. Looks pretty neat otherwise.

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md (hunk 6 above)
Comment (ingvagabund) [reply contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:90]: Are those worth mentioning given they are already part of the mentioned codebase locations? I can add another bullet saying "Regular unit tests ending with `_test.go` located alongside mention codebase locations". Though given unit tests are more or less must for every "serious" implementation, I consider them already included.

File: contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md (hunk 16 above)
Comment (ingvagabund) [reply contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md:264]: Updated

**Comment by Huang-Wei** [comment]
LGTM. Please help squash the commits.

**Comment by alculquicondor** [comment]
+1
Let's merge and iterate :)

**Comment by ingvagabund** [comment]
Squashed
